### ISSUE TITLE
[RFC][HttpKernel] Configurable argument value resolvers

### ISSUE THREAD

**Issue by vudaltsov** [symfony/symfony#29692]
Hi everyone!

Recently got interested in `@ParamConverter`s as a convenient extension point in Symfony. Some time ago in https://github.com/sensiolabs/SensioFrameworkExtraBundle/pull/436 @fabpot proposed an idea on how to implement configurable argument resolvers. Let me explain the idea:
- A single annotation `@Arg(string $name, array options = [])` is proposed as a replacement for `@ParamConverter`.
- A new interface `ArgumentValueOptionsInterface extends ArgumentValueResolverInterface` adds `configure(OptionsResolver $resolver)` method.
- A new subscriber `ControllerArgumentOptionsListener` listens to the `KernelEvents::CONTROLLER` event and looks for `@Arg` mappings for the current controller callable. Then it iterates over the `ArgumentValueOptionsInterface[]` instances to find a resolver based on `supports(Request $request, ArgumentMetadata $argument)`. Once the resolver is found, `@Arg.options` are resolved and saved to `$request->attributes->set($name.'_options', $resolvedOptions)`.
- Later in `SomeArgumentValueResolver::resolve(Request $request, ArgumentMetadata $argument)` we receive resolved options by calling `$request->attributes->get($name.'_options', [])` and use them to yield the value.

While I think that this is a nice solution (especially when looking from the perspective of the `FrameworkExtraBundle`), it has one important drawback. `@ParamConverter` has a possibility to specify converter explicitly by leveraging the `converter` argument. It allows to resolve arguments which by themselves do not provide enough information to find a resolver. A great example is `@ParamConverter("name", converter="fos_rest.request_body")` from [FOSRestBundle](https://symfony.com/doc/master/bundles/FOSRestBundle/request_body_converter_listener.html). This converter allows to deserialize `$request->getContent()` into a DTO based on the argument type hint. Unlike `SomeEntityClass` which we can check with `ManagerRegistry::getManagerForClass()`, `SomeDTOClass` does not give any hint that it should be processed by a `RequestBodyParamConverter`. So, the `@Arg` concept without a possibility to be explicit about the resolver seems to be not suitable for this case.

I have a different idea. It think, it fits the `HttpKernel` much better than the bundle, that's why I am creating an issue here. 

_Btw, the idea to move `@ParamConverters` configuration to the core was already discussed in https://github.com/sensiolabs/SensioFrameworkExtraBundle/pull/436#issuecomment-249283532._

### Design

1. Let's agree that once the argument is explicitly configured (for example by an annotation), we already know what resolver it should be processed by. It's like with validator constraints. So I propose to create the following interface:

    ```php
    interface ArgumentConfigInterface
    {
        /**
         * Returns the FQCN of the argument value resolver that resolves the argument with this config.
         */
        public function resolvedBy(): string;
    }
    ```

1. I think that within the `HttpKernel` component this config might be a part of argument's metadata. Later it will allow us to cache everything together easily. Thus let's add another property to the `ArgumentMetadata`:

    ```php
    namespace Symfony\Component\HttpKernel\ControllerMetadata;

    class ArgumentMetadata
    {
        // ...

        private $config;

        public function __construct(
            string $name,
            ?string $type,
            bool $isVariadic,
            bool $hasDefaultValue,
            $defaultValue,
            bool $isNullable = false,
            ?ArgumentConfigInterface $config = null
        ) {
           // ...

           $this->config = $config;
        }

        // ...

        public function getConfig(): ?ArgumentConfigInterface
        {
            return $this->config;
        }
    }
    ```

1. Should we stick to the annotation mapping? No, let's create an extension point!

    ```php
    interface ArgumentConfigLoaderInterface
    {
        public function getArgumentConfig(callable $controller, string $name): ?ArgumentConfigInterface;
    }

    final class AnnotationArgumentConfigLoader implements ArgumentConfigLoaderInterface
    {
        public function __construct(Doctrine\Common\Annotations\Reader $reader)
        {
            // ...
        }
    }
    ```

    Now we can inject `ArgumentConfigLoaderInterface` into `ArgumentMetadataFactory` and produce config aware `ArgumentMetadata` instances.

1. Next, let's adjust `ArgumentResolver`. Previously we retrieved converters only by iteration, now we want to get them by FQCN as well.

    ```php
    namespace Symfony\Component\HttpKernel\Controller;

    final class ArgumentResolver implements ArgumentResolverInterface
    {
        // ...

        private $argumentValueResolversByClass;

        public function __construct(
            // ...
            ?ContainerInterface $argumentValueResolversByClass = null
        ) {
            // ...
            $this->argumentValueResolversByClass = $argumentValueResolversByClass ?: self::getDefaultArgumentValueResolversByClass();
        }

        // ...

        private function getResolver(Request $request, ArgumentMetadata $metadata): ArgumentValueResolverInterface
        {
            if (null !== $config = $metadata->getConfig()) {
                return $this->argumentValueResolversByClass->get($config->resolvedBy());
            }

            foreach ($this->argumentValueResolvers as $resolver) {
                if ($resolver->supports($request, $metadata)) {
                    return $resolver;
                }
            }

            throw new \RuntimeException();
        }
    }
    ```

1. I think we should also cache all argument metadata. That can be easily done with a decorator `CachingArgumentMetadataFactory implements ArgumentMetadataFactoryInterface`.

After all these non-breaking BC steps we'll be able to create configs like `@Entity`, `@BodyConverter`, `@GetParam` with ease.

### Benefits

- Custom configs for argument value resolvers.
- Faster argument value resolver resolution for configured arguments.
- One point to cache everything.
- Out-of-the-box improved `@ParamConverters` with no need to require `sensio/framework-extra-bundle`.

Ready to discuss!
ping @fabpot, @iltar.

**Comment by javiereguiluz** [comment]
Valentin, thanks for proposing this improvement. In order to have more arguments to decide in favor or against this proposal, we'd need to also know the "drawbacks". You only listed the "benefits".

Maybe it doesn't have any important drawback, but please double check it for things like: performance impact, app upgrade pain, incompatibilities with other Symfony features, etc.

**Comment by vudaltsov** [comment]
@javiereguiluz , thanx for editing the title, now I know the difference between WIP and RFC 😂

### Performance impact

It's not going to be slower than `@ParamConverter`s and will be definitely faster if we add caching for `ArgumentMetadata`

### Upgrade

My suggestion is to implement this in `HttpKernel`, so it's a new feature for Symfony, not for FrameworkExtraBundle. It might be disabled by default in the `FrameworkBundle` config. After we implement `@Entity` in `DoctrineBridge`, one can just change the annotation and switch configs.

**Comment by iltar** [comment]
In your `ArgumentConfigInterface` you have defined a `resolvedBy` method. This is then called as `return $this->argumentValueResolversByClass->get($config->resolvedBy());`. What's the idea you have behind the value `resolvedBy` returns? Will this be a custom class you implement? Do you have an example of this?

**Comment by vudaltsov** [comment]
@iltar , the `ArgumentConfigInterface::resolvedBy` method returns the string class name of the exact argument resolver.

Consider a controller:

```php
use App\Entity\Product;

final class ProductController
{
    /**
     * @Route("/product/{id<\d+>}")
     */
    public function __invoke(Product $product): Response
    {
    }
}
```

In this case we don't know what `Product $product` argument really is. That is why we ask each `ArgumentValueResolverInterface` whether it supports `Product $product` or not. When `EntityValueResolver` says "yes, I know how to handle this" by returning `true` in the `supports` method, we call `$resolver->resolve()`.

```php
use App\Entity\Product;

final class ProductController
{
    /**
     * @Route("/product/{id<\d+>}")
     * @Entity("product", expr="repository.findById(id)")
     */
    public function __invoke(Product $product): Response
    {
    }
}
```
In this case we don't need to iterate over value resolvers. What for? We already know that `$product` should be handled by `EntityValueResolver`. We just fetch this resolver by name obtained from `ArgumentConfigInterface::resolvedBy` and call `resolve`. As I said earlier, when mapping is present, it's like with validator constraints: nothing to guess.

This mechanism will also solve a common `@Entity User implements UserInterface` argument problem. Once we put `@CurrentUser("user")` annotation, the resolver will be `Symfony\Component\Security\Http\Controller\UserValueResolver` and not `EntityValueResolver`.

**Comment by vudaltsov** [comment]
I can make an example of my idea in a simple `website-skeleton` project with some `HttpKernel` services replaced.

**Comment by iltar** [comment]
So if I understand you correctly (and to formulate the feature in a short sentence), you want to pre-bind a resolver to a specific variable name based on config (which can be annotations, xml, yaml etc).

**Comment by vudaltsov** [comment]
Yes, exactly. What do you think?

**Comment by iltar** [comment]
I think it's a great idea, as you probably already figured from the other issue you linked.

**Comment by curry684** [comment]
Never understood why this wasn't native functionality to begin with, it's awesome once you start using and extending the mechanisms.

**Comment by vudaltsov** [comment]
It looks like a lot of people are interested in this and everyone is positive about the implementation idea. May I start working on a PR?

**Comment by adrenalinkin** [comment]
Hello everyone! Past year i was facing the problem which looks same to described but relates to Swagger Documentation.
I tried to find the way to validate request data according to the generated documentation. But i did not found any realization for that.
So I started to realize solution for my purpose. When basic possibility has been achievement by [swagger-resolver-bundle](https://github.com/adrenalinkin/swagger-resolver-bundle) and validation has been provided I am did something very same to described behavior like in the current PR ([realization example](https://gist.github.com/adrenalinkin/d9bfed953034361b3f60030a35919ac0)).

So I thought about resolving any Request (not only described by swagger doc). And @vudaltsov want to offer something like that by current PR. 

I vote YES! And open to help with realization if that will be need.
